libass/ass_utils.c

```diff
@@ -56,6 +56,14 @@ char *ass_strndup_fallback(const char *s, size_t n)
 }
 #endif
 
+#ifndef HAVE_STRCHRNUL
+char *ass_strchrnul_fallback(const char *s, int c)
+{
+    char *pos = strchr(s, c);
+    return pos ? pos : (char*)s + strlen(s);
+}
```

Comment:
This will _regress_ performance for some platforms. With `strchr` + `strlen` this will now need two passes over the string if no match was found.

Unterminated drawings are ime common, so this might become relevant for long drawings, i.e. exactly the case this improves (or at least doesn't meaningfully regress) if `strchrnul` is available.  
Given if both variants are vectorised by the standard library this might still end up faster. E.g. in binonic libc none of the three relevant `str*` is vectorised (but bionic has `strchrnul`) and it’s file layout suggests its non-vectorised `memchr` (used by strlen) originates from OpenBSD, suggesting there might (have been) no vectorisation there either and OpenBSD lacks `strchrnul`.

Ofc the second-next commit replaces this, and completely gets ridd of those two calls but given this is likely to held back this seems noteworthy.
I’m not sure if there’s any better approach than explicitly limiting this to known-vectorised implementations and using a simple loop otherwise. But even this seems kinda pointless if we’ll end up replacing the function calls on all vectorised platforms with bespoke single-pass vectorised code. 

I think it may be better to just use a single, simple loop here for now.

(Note: the only platforms i know lack `strchrnul` but probably have vectorised, `str(len|chr)` are MacOS and Windows, are there any more? In particular any platforms where standard lib functions might be optimised but bespoke vectorisation with overread may be unsafe?)

Reply:
I did my perf testing for this on macOS, where `strchr` + `strlen` is around 7 times faster than the current code, with a file that's composed almost entirely of unterminated drawings. The vectorized `strchrnul` loop is about twice as fast as strchr+strlen, but has a smaller impact on total runtime since by that point `strchrnul` isn't eating up anywhere near as large of a portion of our cycles anymore.

On Windows, as far as I can tell, UCRT has an all-assembly strlen implementation for i386 and amd64 that looks almost identical to my `size_t` implementation here. Not sure about strchr, and MSVCRT appears to only have trivial scalar implementations (like bionic). MinGW has [this implementation](https://github.com/mingw-w64/mingw-w64/blob/819a6ec2ea87c19814b287e21d65e0dc7f05abba/mingw-w64-crt/string/strchrnul.c#L35-L51), which is very similar to musl's, but it's behind a `__GNUC__` check, and uses `__attribute__((__may_alias__))` instead of `memcpy`. UCRT is the only implementation I have source for that uses assembly instead of C for this.

It sounds like what we want here is:
- If we know it to be safe (and have an implementation for the platform), use the 16-byte-wide extension-based vectorized code
- Otherwise, if we have `strchrnul`, use that
- Otherwise, if we know it to be safe (which should be on all the platforms we're discussing here), use the size_t-wide standard-C vectorized code
- Otherwise, use a plain C one-byte-at-a-time loop

This means we'll want this to be `ass_strchrnul` instead of a fallback, and should ensure that this doesn't cause a regression anywhere.

Also, should we see if we're using `strlen` or `strchr` in other hot-path code, and consider using our own implementations of those when possible as well? It sounds like this might actually be a problem on Android.

Re: @MrSmile's comment, as discussed on IRC, I don't think there's any reason to expect issues when using `memcpy` with this in real usage (or perhaps `__attribute__((__may_alias__))` instead). I prefer not to use inline assembly here, since it means much more arch-specific code and more complexity when porting this to other platforms in the future. With the `size_t` implementation, we don't need any arch- or compiler-specific code at all (beyond possibly an ifdef for whether or not we think it's safe to use). Using out-of-line assembly provides very strong safety guarantees (ironically enough), but makes checking for CPU feature support more complex; in the current code, I'm simply using SSE2 or NEON when available (i.e. the compiler was run with `-msse2`, or a platform that always has the extension was used).